File: Plugins/iOS/CrashGuardBridge.swift
import Foundation
import CrashGuardSDK

// MARK: - Callback Types

public typealias StringCallback = @convention(c) (UnsafePointer<CChar>?) -> Void
public typealias BoolCallback = @convention(c) (Bool) -> Void

// MARK: - Helper to convert String to C string

private func withCString(_ string: String?, callback: StringCallback?) {
    guard let callback = callback else { return }
    if let string = string {
        string.withCString { ptr in
            callback(ptr)
        }
    } else {
        callback(nil)
    }
}

private func toCString(_ string: String?) -> UnsafeMutablePointer<CChar>? {
    guard let string = string else { return nil }
    let count = string.utf8.count + 1
    let result = UnsafeMutablePointer<CChar>.allocate(capacity: count)
    string.withCString { ptr in
        result.initialize(from: ptr, count: count)
    }
    return result
}

// MARK: - Initialization

@_cdecl("_CrashGuard_Start")
public func _CrashGuard_Start() {
    CrashGuard.start()
}

@_cdecl("_CrashGuard_IsInitialized")
public func _CrashGuard_IsInitialized() -> Bool {
    return true // SDK auto-initializes on shared access
}

// MARK: - Stability Analysis

@_cdecl("_CrashGuard_AnalyzeStability")
public func _CrashGuard_AnalyzeStability(_ callback: @escaping StringCallback) {
    CrashGuard.shared.analyzeStability { result in
        switch result {
        case .success(let analysis):
            let json = """
            {"id":"\(analysis.id)","signature":"\(analysis.signature)","requiresAttention":\(analysis.requiresAttention),"isActiveDiagnostics":\(analysis.isActiveDiagnostics),"isDebugBuild":\(analysis.isDebugBuild ?? false)}
            """
            json.withCString { ptr in
                callback(ptr)
            }
        case .failure(let error):
            let json = """
            {"error":"\(error.localizedDescription)"}
            """
            json.withCString { ptr in
                callback(ptr)
            }
        }
    }
}

@_cdecl("_CrashGuard_RequiresAttention")
public func _CrashGuard_RequiresAttention(_ callback: @escaping BoolCallback) {
    CrashGuard.shared.requiresAttention { result in
        switch result {
        case .success(let requires):
            callback(requires)
        case .failure:
            callback(false)
        }
    }
}

@_cdecl("_CrashGuard_RequiresAttentionCached")
public func _CrashGuard_RequiresAttentionCached() -> Bool {
    return CrashGuard.shared.lastAnalysis?.requiresAttention ?? false
}

@_cdecl("_CrashGuard_VerifySignature")
public func _CrashGuard_VerifySignature(_ signature: UnsafePointer<CChar>?, _ callback: @escaping StringCallback) {
    guard let signature = signature else {
        callback(nil)
        return
    }
    let signatureStr = String(cString: signature)

    CrashGuard.shared.verifySignature(signatureStr) { result in
        switch result {
        case .success(let verification):
            let json = """
            {"signature":"\(verification.signature)","isFlagged":\(verification.isFlagged)}
            """
            json.withCString { ptr in
                callback(ptr)
            }
        case .failure(let error):
            let json = """
            {"error":"\(error.localizedDescription)"}
            """
            json.withCString { ptr in
                callback(ptr)
            }
        }
    }
}

// MARK: - Runtime Info

@_cdecl("_CrashGuard_GetRuntimeInfo")
public func _CrashGuard_GetRuntimeInfo(_ callback: @escaping StringCallback) {
    let info = CrashGuard.shared.runtimeInfo
    let json = """
    {"isDebugEnvironment":\(info.isDebugEnvironment),"isBetaBuild":\(info.isBetaBuild),"isEmulated":\(info.isEmulated),"hasRuntimeModifications":\(info.hasRuntimeModifications),"hasDiagnosticsEnabled":\(info.hasDiagnosticsEnabled)}
    """
    json.withCString { ptr in
        callback(ptr)
    }
}

// MARK: - Quick Checks

@_cdecl("_CrashGuard_IsDebugEnvironment")
public func _CrashGuard_IsDebugEnvironment() -> Bool {
    return CrashGuard.shared.isDebugEnvironment()
}

@_cdecl("_CrashGuard_IsBetaBuild")
public func _CrashGuard_IsBetaBuild() -> Bool {
    return CrashGuard.shared.isBetaBuild()
}

@_cdecl("_CrashGuard_HasRuntimeModifications")
public func _CrashGuard_HasRuntimeModifications() -> Bool {
    return CrashGuard.shared.hasRuntimeModifications()
}

@_cdecl("_CrashGuard_IsEmulated")
public func _CrashGuard_IsEmulated() -> Bool {
    return CrashGuard.shared.runtimeInfo.isEmulated
}

// MARK: - Device Identifiers

@_cdecl("_CrashGuard_GetPersistentId")
public func _CrashGuard_GetPersistentId() -> UnsafeMutablePointer<CChar>? {
    return toCString(CrashGuard.shared.persistentId)
}

@_cdecl("_CrashGuard_GetVendorId")
public func _CrashGuard_GetVendorId() -> UnsafeMutablePointer<CChar>? {
    return toCString(CrashGuard.shared.vendorId)
}

@_cdecl("_CrashGuard_GetCrashSignature")
public func _CrashGuard_GetCrashSignature() -> UnsafeMutablePointer<CChar>? {
    return toCString(CrashGuard.shared.crashSignature)
}

// MARK: - Event Tracking

@_cdecl("_CrashGuard_RecordScreenTransition")
public func _CrashGuard_RecordScreenTransition(_ to: UnsafePointer<CChar>?, _ from: UnsafePointer<CChar>?) {
    guard let to = to else { return }
    let toStr = String(cString: to)
    let fromStr = from != nil ? String(cString: from!) : nil
    CrashGuard.shared.recordScreenTransition(to: toStr, from: fromStr)
}

@_cdecl("_CrashGuard_RecordTransactionView")
public func _CrashGuard_RecordTransactionView(_ screen: UnsafePointer<CChar>?) {
    let screenStr = screen != nil ? String(cString: screen!) : "payment"
    CrashGuard.shared.recordTransactionView(screen: screenStr)
}

@_cdecl("_CrashGuard_RecordTransactionDismiss")
public func _CrashGuard_RecordTransactionDismiss(_ screen: UnsafePointer<CChar>?) {
    let screenStr = screen != nil ? String(cString: screen!) : "payment"
    CrashGuard.shared.recordTransactionDismiss(screen: screenStr)
}

@_cdecl("_CrashGuard_RecordTransactionAttempt")
public func _CrashGuard_RecordTransactionAttempt(_ screen: UnsafePointer<CChar>?) {
    let screenStr = screen != nil ? String(cString: screen!) : "payment"
    CrashGuard.shared.recordTransactionAttempt(screen: screenStr)
}

@_cdecl("_CrashGuard_RecordTransactionSuccess")
public func _CrashGuard_RecordTransactionSuccess(_ screen: UnsafePointer<CChar>?) {
    let screenStr = screen != nil ? String(cString: screen!) : "payment"
    CrashGuard.shared.recordTransactionSuccess(screen: screenStr)
}

@_cdecl("_CrashGuard_RecordPermissionRequest")
public func _CrashGuard_RecordPermissionRequest(_ permissionType: UnsafePointer<CChar>?) {
    guard let permissionType = permissionType else { return }
    let permissionStr = String(cString: permissionType)
    CrashGuard.shared.recordPermissionRequest(permissionStr)
}

@_cdecl("_CrashGuard_RecordSettingsAccess")
public func _CrashGuard_RecordSettingsAccess(_ screen: UnsafePointer<CChar>?) {
    let screenStr = screen != nil ? String(cString: screen!) : "settings"
    CrashGuard.shared.recordSettingsAccess(screen: screenStr)
}

@_cdecl("_CrashGuard_RecordEvent")
public func _CrashGuard_RecordEvent(_ name: UnsafePointer<CChar>?, _ contextJson: UnsafePointer<CChar>?) {
    guard let name = name else { return }
    let nameStr = String(cString: name)

    var context: [String: Any]? = nil
    if let contextJson = contextJson {
        let contextStr = String(cString: contextJson)
        if let data = contextStr.data(using: .utf8),
           let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            context = json
        }
    }

    CrashGuard.shared.recordEvent(nameStr, context: context)
}

@_cdecl("_CrashGuard_FlushEvents")
public func _CrashGuard_FlushEvents(_ callback: @escaping BoolCallback) {
    CrashGuard.shared.flushEvents { result in
        switch result {
        case .success:
            callback(true)
        case .failure:
            callback(false)
        }
    }
}

// MARK: - Memory Management

@_cdecl("_CrashGuard_FreeString")
public func _CrashGuard_FreeString(_ ptr: UnsafeMutablePointer<CChar>?) {
    ptr?.deallocate()
}

File: Plugins/Android/CrashGuardBridge.java
package com.crashguard.unity;

import android.app.Activity;
import android.util.Log;

import com.crashguard.sdk.CrashGuard;
import com.crashguard.sdk.RuntimeInfo;
import com.crashguard.sdk.network.AnalysisResult;
import com.crashguard.sdk.network.SignatureVerification;
import com.unity3d.player.UnityPlayer;

import org.json.JSONObject;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import kotlin.Unit;
import kotlin.Result;

/**
 * Bridge between Unity C# and CrashGuard Android SDK
 */
public class CrashGuardBridge {
    private static final String TAG = "CrashGuardBridge";

    private static Activity getActivity() {
        return UnityPlayer.currentActivity;
    }

    // ============== Initialization ==============

    public static void start() {
        Activity activity = getActivity();
        if (activity != null) {
            CrashGuard.start(activity);
        }
    }

    public static boolean isInitialized() {
        return CrashGuard.isInitialized();
    }

    // ============== Stability Analysis ==============

    public static void analyzeStability(final String gameObject, final String methodName) {
        try {
            CrashGuard.Companion.getShared().analyzeStability(result -> {
                try {
                    if (result.isSuccess()) {
                        AnalysisResult analysis = result.getOrNull();
                        if (analysis != null) {
                            JSONObject json = new JSONObject();
                            json.put("id", analysis.getId());
                            json.put("signature", analysis.getSignature());
                            json.put("requiresAttention", analysis.getRequiresAttention());
                            json.put("isActiveDiagnostics", analysis.isActiveDiagnostics());
                            json.put("isDebugBuild", analysis.isDebugBuild());
                            UnityPlayer.UnitySendMessage(gameObject, methodName, json.toString());
                        }
                    } else {
                        JSONObject json = new JSONObject();
                        Throwable error = result.exceptionOrNull();
                        json.put("error", error != null ? error.getMessage() : "Unknown error");
                        UnityPlayer.UnitySendMessage(gameObject, methodName, json.toString());
                    }
                } catch (Exception e) {
                    Log.e(TAG, "Error in analyzeStability callback", e);
                }
                return Unit.INSTANCE;
            });
        } catch (Exception e) {
            Log.e(TAG, "Error calling analyzeStability", e);
        }
    }

    public static void requiresAttention(final String gameObject, final String methodName) {
        try {
            CrashGuard.Companion.getShared().requiresAttention(result -> {
                try {
                    Boolean requires = result.isSuccess() ? result.getOrNull() : false;
                    UnityPlayer.UnitySendMessage(gameObject, methodName, requires != null && requires ? "true" : "false");
                } catch (Exception e) {
                    Log.e(TAG, "Error in requiresAttention callback", e);
                }
                return Unit.INSTANCE;
            });
        } catch (Exception e) {
            Log.e(TAG, "Error calling requiresAttention", e);
        }
    }

    public static boolean requiresAttentionCached() {
        try {
            AnalysisResult lastAnalysis = CrashGuard.Companion.getShared().getLastAnalysis();
            return lastAnalysis != null && lastAnalysis.getRequiresAttention();
        } catch (Exception e) {
            return false;
        }
    }

    public static void verifySignature(String signature, final String gameObject, final String methodName) {
        // Note: verifySignature is a suspend function in Kotlin
        // We need to handle this appropriately - for now, send error
        try {
            JSONObject json = new JSONObject();
            json.put("error", "verifySignature not yet implemented for Android Unity bridge");
            UnityPlayer.UnitySendMessage(gameObject, methodName, json.toString());
        } catch (Exception e) {
            Log.e(TAG, "Error in verifySignature", e);
        }
    }

    // ============== Runtime Info ==============

    public static void getRuntimeInfo(final String gameObject, final String methodName) {
        try {
            RuntimeInfo info = CrashGuard.Companion.getShared().getRuntimeInfo();
            JSONObject json = new JSONObject();
            json.put("isDebugEnvironment", info.isDebugEnvironment());
            json.put("isEmulated", info.isEmulated());
            json.put("hasRuntimeModifications", info.getHasRuntimeModifications());
            json.put("isDeveloperMode", info.isDeveloperMode());
            json.put("isAdbEnabled", info.isAdbEnabled());
            json.put("isBetaBuild", false); // Android doesn't have TestFlight equivalent
            json.put("hasDiagnosticsEnabled", info.isDeveloperMode());
            UnityPlayer.UnitySendMessage(gameObject, methodName, json.toString());
        } catch (Exception e) {
            Log.e(TAG, "Error getting runtime info", e);
        }
    }

    // ============== Quick Checks ==============

    public static boolean isDebugEnvironment() {
        try {
            return CrashGuard.Companion.getShared().isDebugEnvironment();
        } catch (Exception e) {
            return false;
        }
    }

    public static boolean isEmulated() {
        try {
            return CrashGuard.Companion.getShared().isEmulated();
        } catch (Exception e) {
            return false;
        }
    }

    public static boolean hasRuntimeModifications() {
        try {
            return CrashGuard.Companion.getShared().hasRuntimeModifications();
        } catch (Exception e) {
            return false;
        }
    }

    public static boolean isDeveloperMode() {
        try {
            return CrashGuard.Companion.getShared().isDeveloperMode();
        } catch (Exception e) {
            return false;
        }
    }

    public static boolean isFromPlayStore() {
        try {
            return CrashGuard.Companion.getShared().isFromPlayStore();
        } catch (Exception e) {
            return false;
        }
    }

    // ============== Device Identifiers ==============

    public static String getPersistentId() {
        try {
            return CrashGuard.Companion.getShared().getPersistentId();
        } catch (Exception e) {
            return "";
        }
    }

    public static String getCrashSignature() {
        try {
            return CrashGuard.Companion.getShared().getCrashSignature();
        } catch (Exception e) {
            return null;
        }
    }

    // ============== Event Tracking ==============

    public static void recordScreenTransition(String to, String from) {
        try {
            CrashGuard.Companion.getShared().recordScreenTransition(to, from);
        } catch (Exception e) {
            Log.e(TAG, "Error recording screen transition", e);
        }
    }

    public static void recordTransactionView(String screen) {
        try {
            CrashGuard.Companion.getShared().recordTransactionView(screen != null ? screen : "payment");
        } catch (Exception e) {
            Log.e(TAG, "Error recording transaction view", e);
        }
    }

    public static void recordTransactionDismiss(String screen) {
        try {
            CrashGuard.Companion.getShared().recordTransactionDismiss(screen != null ? screen : "payment");
        } catch (Exception e) {
            Log.e(TAG, "Error recording transaction dismiss", e);
        }
    }

    public static void recordTransactionAttempt(String screen) {
        try {
            CrashGuard.Companion.getShared().recordTransactionAttempt(screen != null ? screen : "payment");
        } catch (Exception e) {
            Log.e(TAG, "Error recording transaction attempt", e);
        }
    }

    public static void recordTransactionSuccess(String screen) {
        try {
            CrashGuard.Companion.getShared().recordTransactionSuccess(screen != null ? screen : "payment");
        } catch (Exception e) {
            Log.e(TAG, "Error recording transaction success", e);
        }
    }

    public static void recordPermissionRequest(String permissionType) {
        try {
            CrashGuard.Companion.getShared().recordPermissionRequest(permissionType);
        } catch (Exception e) {
            Log.e(TAG, "Error recording permission request", e);
        }
    }

    public static void recordSettingsAccess(String screen) {
        try {
            CrashGuard.Companion.getShared().recordSettingsAccess(screen != null ? screen : "settings");
        } catch (Exception e) {
            Log.e(TAG, "Error recording settings access", e);
        }
    }

    public static void recordEvent(String name, String contextJson) {
        try {
            Map<String, Object> context = null;
            if (contextJson != null && !contextJson.isEmpty()) {
                context = new HashMap<>();
                JSONObject json = new JSONObject(contextJson);
                Iterator<String> keys = json.keys();
                while (keys.hasNext()) {
                    String key = keys.next();
                    context.put(key, json.get(key));
                }
            }
            CrashGuard.Companion.getShared().recordEvent(name, context);
        } catch (Exception e) {
            Log.e(TAG, "Error recording event", e);
        }
    }

    public static void flushEvents(final String gameObject, final String methodName) {
        try {
            CrashGuard.Companion.getShared().flushEvents(result -> {
                UnityPlayer.UnitySendMessage(gameObject, methodName, result.isSuccess() ? "true" : "false");
                return Unit.INSTANCE;
            });
        } catch (Exception e) {
            Log.e(TAG, "Error flushing events", e);
            UnityPlayer.UnitySendMessage(gameObject, methodName, "false");
        }
    }
}
